packages/apps/google-analytics/src/google-analytics-events.ts

```diff
@@ -0,0 +1,181 @@
+
+  const buyer = order.buyers && order.buyers[0];
+  const clientIp = order.client_ip;
+
+  const enabledCustonEvent = order.financial_status?.current && appData.custom_events;
+  const enabledRefundEvent = order.status === 'cancelled' && appData.refund_event !== false;
+
+  if (orderId && buyer && clientIp && order.items) {
+    try {
+      if (measurementId && apiSecret && (enabledCustonEvent || enabledRefundEvent)) {
+        const url = `/mp/collect?api_secret=${apiSecret}&measurement_id=${measurementId}`;
+
+        const items = order.items.map((item) => {
+          const eventItem: EventItem = {
+            item_id: item.product_id,
+            item_name: item.name || item.sku || '',
+            price: item.final_price || item.price,
+            quantity: item.quantity,
+          };
+
+          if (item.variation_id) {
+            eventItem.item_variant = item.variation_id;
+          }
+          if (item.kit_product) {
+            eventItem.item_list_id = item.kit_product._id;
+            eventItem.item_list_name = item.kit_product.name;
+          }
+          return eventItem;
+        });
+
+        const params: EventParams = {
+          id: orderId,
+          currency: order.currency_id || 'BRL',
+          transaction_id: orderId,
+          value: order.amount.total,
+          status: order.financial_status?.current || order.status,
+          items,
+        };
+        if (order.amount.freight) {
+          params.shipping = order.amount.freight;
+        }
+
+        if (order.amount.tax || order.amount.extra) {
+          params.tax = (order.amount.tax || 0) + (order.amount.extra || 0);
+        }
+
+        if (order.extra_discount?.discount_coupon) {
+          params.coupon = order.extra_discount.discount_coupon;
+        }
+
+        const events: {
+          name: string,
+          params: EventParams
+        }[] = [];
+
+        const collectionEventsSent = getFirestore().collection('googleAnalyticsEventsSent');
+        let customEventsSent: string[] = [];
+        const eventFirestore = await getSentEventFirestore(collectionEventsSent, orderId);
+
+        if (enabledCustonEvent && order.financial_status) {
+          const eventName = `purchase_${order.financial_status.current}`;
+          customEventsSent = (eventFirestore && eventFirestore.customEventsSent) || [];
+
+          if (customEventsSent[customEventsSent.length - 1] !== eventName) {
+            events.push({
+              name: eventName,
+              params,
+            });
+
+            customEventsSent.push(eventName);
+          }
+        }
+
+        if (enabledRefundEvent && (!eventFirestore || eventFirestore.status !== 'cancelled')) {
+          events.push({
+            name: 'refund',
+            params,
+          });
+        }
+        // https://developers.google.com/analytics/devguides/collection/ga4/reference/events?client_type=gtag#purchase
+
+        const body = {
+          client_id: buyer._id,
```

Comment:
Que `client_id` é esse? Precisa disso? 👀 
Fora dos payload dos eventos? 🤔

Reply:
E `buyer` pode ser undefined em um pedido manual, esse trecho aqui pode dar fatal

Reply:
Eu acho que pode mandar só `{ events }` mesmo na verdade

Reply:
Vou testar aqui, se funciona sem, mas na doc deles lá o client_id é um parametro obrigatório, porem gera algo ate capturado com gtag no exemplo, mas em alguns testes que fiz funcionou. (na v1) 

Pq na verdade um client id pode enviar mais de um evento para GA4 de uma vezes (pelo que entendi), o payload de exemplo la segue assim: 
```
{
"client_id": "x",
"events": [
  {
    "name": "offline_purchase",
    "params": {
      "engagement_time_msec": "100",
      "session_id": "123"
    }
  }
]
}
```

Reply:
Bom, se for obrigatório você deveria usar algo como `buyer?._id || Date.now().toString()` , sem risco de dar fatal se `buyer` undefined e com um fallback randômico..